### PR TITLE
JDK-8315042 NPE in PKCS7.parseOldSignedData

### PR DESCRIPTION
**Pull request by mcpowers**
https://bugs.openjdk.org/browse/JDK-8315042

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315042](https://bugs.openjdk.org/browse/JDK-8315042): NPE in PKCS7.parseOldSignedData (**Bug** - P3)

### Reviewers
 * [Valerie Peng](https://openjdk.org/census#valeriep) (@valeriepeng - **Reviewer**) ⚠️ Review applies to [2fe7e47c](https://git.openjdk.org/jdk/pull/15844/files/2fe7e47c744c564356f2a31a2eb04b01871eb505)
 * [Weijun Wang](https://openjdk.org/census#weijun) (@wangweij - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15844/head:pull/15844` \
`$ git checkout pull/15844`

Update a local copy of the PR: \
`$ git checkout pull/15844` \
`$ git pull https://git.openjdk.org/jdk.git pull/15844/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15844`

View PR using the GUI difftool: \
`$ git pr show -t 15844`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15844.diff">https://git.openjdk.org/jdk/pull/15844.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15844#issuecomment-1727922839)

### CODE REVIEW

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2004, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2004, 2023 Oracle and/or its affiliates. All rights reserved.
```
Comment (valeriepeng) [inline test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:2]: missing a coma after 2023.

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java
[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:1]: I can do that. My primary motivation for using the existing test was that `UnexpectedNPE.java` is the perfect name for my test.

[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:1]: Why not just updating UnexpectedNPE with `Utils.runAndCheckException`?

[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:1]: And leave `Utils.runAndCheckException` out of the fix? I'm fine either way.

[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:1]: +1.

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java [hunk 2]
```diff
@@ -65,14 +64,13 @@ private boolean run(byte[] buf) {
...
         }
-        try {
-            cf.generateCRL(new ByteArrayInputStream(buf));
-        } catch (CRLException ce) {
-            System.out.println("NPE checking passed");
-            return true;
-        }
 
-        System.out.println("CRLException has not been thrown");
-        return false;
```
[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:76]: I don't understand.

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java [hunk 3]
```diff
@@ -23,39 +23,38 @@
...
 
     public UnexpectedNPE() {}
 
     public static void main( String[] av ) {
         byte[] encoded_1 = { 0x00, 0x00, 0x00, 0x00 };
         byte[] encoded_2 = { 0x30, 0x01, 0x00, 0x00 };
         byte[] encoded_3 = { 0x30, 0x01, 0x00 };
+        byte[] decodedBytes = Base64.getDecoder().decode(in);
 
         UnexpectedNPE unpe = new UnexpectedNPE() ;
```
[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:49]: Good suggestion.

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java [hunk 4]
```diff
@@ -23,39 +23,38 @@
...
 import java.io.ByteArrayInputStream;
+import java.util.Base64;
+
+import jdk.test.lib.Utils;
 
 public class UnexpectedNPE {
     CertificateFactory cf = null ;
+    private static final String in = "MAsGCSqGSMP7TQEHAjI1Bgn///////8wCwUyAQ==";
 
     public UnexpectedNPE() {}
```
[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:41]: Removed.

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java [hunk 5]
```diff
@@ -23,39 +23,38 @@
...
 
     public static void main( String[] av ) {
         byte[] encoded_1 = { 0x00, 0x00, 0x00, 0x00 };
         byte[] encoded_2 = { 0x30, 0x01, 0x00, 0x00 };
         byte[] encoded_3 = { 0x30, 0x01, 0x00 };
+        byte[] decodedBytes = Base64.getDecoder().decode(in);
```
[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:47]: Done.

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java [hunk 6]
```diff
@@ -23,39 +23,38 @@
...
+        unpe.run(encoded_2);
+        unpe.run(encoded_3);
+        unpe.run(decodedBytes);
     }
 
-    private boolean run(byte[] buf) {
+    private void run(byte[] buf) {
         if (cf == null) {
             try {
                 cf = CertificateFactory.getInstance("X.509", "SUN");
```
[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:60]: Doing the former.

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:76]: Fixed.

**Comment by mcpowers** [comment]
@wangweij Thanks for the review - many good comments.

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java [hunk 7]
```diff
@@ -23,39 +23,27 @@
...
-            throw new SecurityException("CRLException has not been thrown");
-        }
-
-        if(!unpe.run(encoded_2)) {
-            throw new SecurityException("CRLException has not been thrown");
-        }
-    }
-
-    private boolean run(byte[] buf) {
         if (cf == null) {
```
[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:47]: Fixed.

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java [hunk 8]
```diff
@@ -23,39 +23,27 @@
 
 /*
  * @test
- * @bug 5052433
+ * @bug 5052433 8315042
  * @summary NullPointerException for generateCRL and generateCRLs methods.
```
[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:27]: I'll revert the code to check both.

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java [hunk 9]
```diff
@@ -65,14 +53,17 @@ private boolean run(byte[] buf) {
...
+        run(encoded_2);
+        run(encoded_3);
+        run(encoded_4);
+    }
+
+    private static void run(byte[] buf) {
+        Utils.runAndCheckException(
+                () -> cf.generateCRLs(new ByteArrayInputStream(buf)),
+                CRLException.class);
+        System.out.println("NPE checking passed");
```
[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:67]: Less is better in my opinion.

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:27]: If we are fixing the summary, can we also elaborate a little more so it's clear that the test is checking for CRLException?

**Review by valeriepeng — APPROVED** [review]

File: test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java:27]: What about:
"Verify that generateCRL and generateCRLs methods do not throw a NullPointerException. They should throw a CRLException instead."

[the remark this answers is not in the document]
Comment (mcpowers) [reply test/jdk/sun/security/x509/X509CRLImpl/UnexpectedNPE.java]: @valeriepeng I improved the summary a bit.

**Comment by mcpowers** [comment]
I changed the summary to be more clear. Can everyone approve again? This will be my last change for sure.

**Pull request closed without merging**